### PR TITLE
add return-to-declaration functionality on the tsserver branch

### PR DESCRIPTION
**Pull request by deanproxy**
[ ] All compiled assets are included (atom packages are git tags and hence the built files need to be a part of the source control)

### CODE REVIEW

File: lib/main/atom/commands/goToDeclaration.ts [hunk 1]
```diff
@@ -2,6 +2,15 @@ import {commands} from "./registry"
 import {commandForTypeScript, getFilePathPosition} from "../utils"
 import {simpleSelectionView} from "../views/simpleSelectionView"
 
+const prevCursorPositions:any[] = [];
```
Comment (guncha) [inline lib/main/atom/commands/goToDeclaration.ts]: Missing the type annotation here. Once you add it, you'll discover a type mismatch between what `getFilePathPosition` returns and what `open` expects.

File: lib/main/atom/commands/goToDeclaration.ts [hunk 2]
```diff
@@ -12,6 +21,8 @@ commands.set("typescript:go-to-declaration", deps => {
     const client = await deps.getClient(location.file)
     const result = await client.executeDefinition(location)
 
+    prevCursorPositions.push(location);
```
Comment (guncha) [inline lib/main/atom/commands/goToDeclaration.ts]: It's possible that the user executes this command, gets multiple results, but presses escape and doesn't open any of them. We don't want to record the previous position in that case - only if they make the jump.

File: lib/main/atom/commands/goToDeclaration.ts [hunk 3]
```diff
@@ -27,12 +38,19 @@ commands.set("typescript:go-to-declaration", deps => {
...
+   return async e => {
+      const position = prevCursorPositions.pop();
+      if (!position) {
+         atom.notifications.addInfo('AtomTS: Previous position not found.');
+         return;
+      }
+      open({
+         file: position.file,
+         start: { line: position.line, offset: position.offset }
+      });
```
Comment (guncha) [inline lib/main/atom/commands/goToDeclaration.ts]: The new object you're creating here has the exact same shape as `position` local, so there's no need for that. You could pass it directly, though you'll see that it's not what `open` expects.

Comment (deanproxy) [inline lib/main/atom/commands/goToDeclaration.ts]: Not completely sure what you mean here. position doesn't have a `start` attribute, it looks like:

    {
        file: string,
        line: number,
        offset: number
    }

Open expect the `line` and `offset` to be children of `start`, which is why I wrote it out like that.  Are you saying to just change the definition of `open` to correspond with the `position` argument?

Comment (guncha) [inline lib/main/atom/commands/goToDeclaration.ts]: Oh, gotcha, I didn't see the `start` bit. Carry on 😆

**Comment by deanproxy** [comment]
added those suggested fixes. let me know if you see anything else.

**Comment by deanproxy** [comment]
Just thinking about this... `prevCursorPosition` array can wind up growing too large and consuming memory if you don't return from your go-to's frequently and you're also one of those people like me that keeps their editor open all of the time. This isn't likely to be a big problem in most situations, but something to keep in mind.

**Comment by guncha** [comment]
Thanks @deanproxy 👍

**Pull request merged**